client/src/partials/vouchers/complex.js

```diff
@@ -0,0 +1,261 @@
+ */
+function ComplexJournalVoucherController(Vouchers, $translate, Accounts, Currencies, Session, FindEntity, FindReference, Notify) {
+  var vm = this;
+
+  // bread crumb paths
+  vm.paths = [{
+    label : $translate.instant('VOUCHERS.COMPLEX.TITLE'),
+    current: true
+  }];
+
+  // bind the startup method as a reset method
+  vm.submit = submit;
+  vm.currencySymbol     = currencySymbol;
+  vm.addVoucherItem     = addVoucherItem;
+  vm.removeVoucherItem  = removeVoucherItem;
+  vm.checkRowValidity   = checkRowValidity;
+  vm.selectAccount      = selectAccount;
+  vm.openEntityModal    = FindEntity.openModal;
+  vm.openReferenceModal = FindReference.openModal;
+
+  // load the list of accounts
+  Accounts.read()
+  .then(function (accounts) {
+    vm.accounts = accounts.map(function (item) {
+      // remove children accounts
+      delete item.children;
+      return item;
+    });
+  });
+
+  // load the available currencies
+  Currencies.read()
+  .then(function (currencies) {
+    currencies.forEach(function (currency) {
+      currency.label = Currencies.format(currency.id);
+    });
+    vm.currencies = currencies;
+  });
+
+  /** Get the selected currency symbol */
+  function currencySymbol(currency_id) {
+    if (!currency_id) { return ; }
+    return Currencies.symbol(currency_id);
+  }
+
+  /** Add transaction row */
+  function addVoucherItem() {
+    vm.rows.push(generateRow());
+    checkRowValidity();
+  }
+
+  /** remove transaction row */
+  function removeVoucherItem(index) {
+    vm.rows.splice(index, 1);
+    checkRowValidity();
+  }
+
+  /** generate row element */
+  function generateRow() {
+    return {
+      account_id    : undefined,
+      debit         : 0,
+      credit        : 0,
+      document_uuid : undefined,
+      entity_uuid   : undefined
+    };
+  }
+
+  /** clean and generate voucher correct data */
+  function handleVoucher() {
+    var voucher;
+    if (vm.sumCredit === vm.sumDebit) {
+      voucher = {
+        project_id : Session.project.id,
+        date : vm.voucher.date,
+        description : vm.voucher.description,
+        currency_id : vm.voucher.currency_id,
+        amount : vm.sumDebit,
+        user_id : Session.user.id,
+      };
+    }
+    return voucher;
+  }
+
+  /** clean and generate voucher items data */
+  function handleVoucherItems() {
+    var voucherItems = [];
+    if (vm.validInput) {
+
+      var entity_uuid = undefined;
+      var document_uuid = undefined;
+
+      vm.rows.forEach(function (row) {
+
+        entity_uuid = row.entity && row.entity.uuid ? row.entity.uuid : '';
+        document_uuid = row.reference && row.reference.document_uuid ? row.reference.document_uuid : '';
+
+        var line = {
+          account_id    : row.account_id,
+          debit         : row.debit,
+          credit        : row.credit,
+          document_uuid : document_uuid,
+          entity_uuid   : entity_uuid
+        };
+
+        voucherItems.push(line);
```

Comment:
Every time you write the following code...
```js
var items = [];

object.otherItems.forEach(function (item) {
  // do some processing

  items.push(item);
});
```
... you are actually doing an `Array.map()` manually.  It is much easier to read if you use `Array.map()`.

Please use a `map()` here.  The code will look something like this:
```js
function handleVoucherItems() {
  return vm.rows.map(function (row) {
    var entity_uuid = row.entity && row.entity.uuid ? row.entity.uuid : null;
    var document_uuid = row.reference && row.reference.document_uuid ? row.reference.document_uuid : null;
    return {
       account_id : row.account_id,
       // ... etc ...
       entity_uuid : entity_uuid
    };
  });
}
```